### PR TITLE
Reduce allocations around empty ReadOnlyCollections

### PR DESCRIPTION
**Pull request by stephentoub**
- Add an internal (for now) `ReadOnlyCollection<T>.Empty`
- Return the singleton `ReadOnlyCollection<T>.Empty` from `Array.AsReadOnly` when the source is empty
- Return a singleton empty enumerator from an empty `ReadOnlyCollection<T>`
- Fix up some reflection code to avoid unnecessary `ReadOnlyCollection<T>` allocations (e.g. just to wrap empty arrays)
- Fix a null reference bug in RuntimeCustomAttributeData.NamedArguments (I don't know if it's actually reachable, but this property shouldn't be returning null)
- Avoid a few unnecessary allocations in TimeZoneInfo.GetSystemTimeZones

Contributes to https://github.com/dotnet/runtime/issues/76028

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs [hunk 1]
```diff
@@ -23,6 +25,25 @@ public ReadOnlyCollection(IList<T> list)
             this.list = list;
         }
 
+        /// <summary>Lazily-initialized empty singleton.</summary>
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs:28]: There is no other static in this type. You can just depend on the beforefieldinit static constructor to initialize this lazily.

File: src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs [hunk 2]
```diff
@@ -37,10 +58,10 @@ public void CopyTo(T[] array, int index)
             list.CopyTo(array, index);
         }
 
-        public IEnumerator<T> GetEnumerator()
-        {
-            return list.GetEnumerator();
-        }
+        public IEnumerator<T> GetEnumerator() =>
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs:61]: Is there a chance that the calling code depends on the exact type of the enumerator returned?

File: src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs:28]: ok

File: src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs:61]: I'm sure someone somewhere is, but I expect it's a very rare case, and it's something we say is unsupported. I can mark it as breaking if you think it's worthwhile.

**Comment by stephentoub** [comment]
> That would be a breaking change for code like

Yes, that's why it doesn't. I wrote the description before I actually implemented things, thought I removed that part of the description, but apparently didn't.

File: src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs (hunk 2 above)
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs:61]: I do not think we need to mark this as breaking. I just wanted to highlight as a potential break.

File: src/libraries/Common/tests/System/Collections/IEnumerable.Generic.Tests.cs [hunk 3]
```diff
@@ -62,6 +62,18 @@ public abstract partial class IEnumerable_Generic_Tests<T> : TestBase<T>
...
+        /// When calling Current of the empty enumerator before the first MoveNext, after the end of the collection,
+        /// or after modification of the enumeration, the resulting behavior is undefined. Tests are included
+        /// to cover two behavioral scenarios:
+        ///   - Throwing an InvalidOperationException
+        ///   - Returning an undefined value.
+        ///
+        /// If this property is set to true, the tests ensure that the exception is thrown. The default value is
+        /// false.
+        /// </summary>
+        protected virtual bool Enumerator_Empty_Current_UndefinedOperation_Throws => Enumerator_Current_UndefinedOperation_Throws;
```
Comment (eiriktsarpalis) [inline src/libraries/Common/tests/System/Collections/IEnumerable.Generic.Tests.cs:75]: Does this need to return the value of `Enumerator_Current_UndefinedOperation_Throws` by default? If so, then the statement in the xml docs above that "the default value is false" is not entirely accurate.
```suggestion
        /// If this property is set to true, the tests ensure that the exception is thrown. The default value is
        /// false.
        /// </summary>
        protected virtual bool Enumerator_Empty_Current_UndefinedOperation_Throws => false;
```

**Review by eiriktsarpalis — APPROVED** [review]

File: src/libraries/Common/tests/System/Collections/IDictionary.Generic.Tests.cs [hunk 4]
```diff
@@ -411,10 +411,12 @@ public void IDictionary_Generic_Keys_Enumeration_ParentDictionaryModifiedInvalid
             }
             else
             {
-                keysEnum.MoveNext();
+                if (keysEnum.MoveNext())
+                {
+                    _ = keysEnum.Current;
```
Comment (krwq) [inline src/libraries/Common/tests/System/Collections/IDictionary.Generic.Tests.cs:416]: I assume this is to validate this doesn't throw?

**Review by krwq — APPROVED** [review]
LGTM, thanks!

File: src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs [hunk 5]
```diff
@@ -2097,15 +2097,23 @@ private ReadOnlyCollection<TKey> GetKeys()
                     ThrowHelper.ThrowOutOfMemoryException();
                 }
 
-                var keys = new List<TKey>(count);
+                if (count == 0)
+                {
+                    // TODO https://github.com/dotnet/runtime/pull/76097: Replace with ReadOnlyCollection<TKey>.Empty.
```
Comment (krwq) [inline src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs:2102]: This TODO is pointing to this PR, is that intended?

Comment (stephentoub) [reply src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs:2102]: It should have been https://github.com/dotnet/runtime/issues/76028.

**Pull request merged**